### PR TITLE
feature_service: disable `GROUP0_SCHEMA_VERSIONING`

### PR DESCRIPTION
**Pull request by kbr-scylla**
It turns out that 3d4398d1b2b9bc2139718e49a806214171ee081f introduced a regression. The dtest `test_add_udt_to_another_data_types` started failing in a flaky way.  The test is adding a field to the user-defined type `user_refs` and then inserting to a table that references this type. The insert fails with the following error: `Invalid tuple literal for entity_info: component 5 is not of type frozen<user_refs>`. But Scylla apparently keeps using the old version of this type, without the new field. And this doesn't seem a race in schema propagation across shards in Scylla -- once the first insert fails, sleeping and retrying the insert fails in the same way, so it seems that schema enters a broken state that is not transient.

The feature wasn't released in any version of Scylla yet, so we can safely disable it for now. When the feature is disabled, Scylla uses the old way of determining schema versions (calculating hashes). I verified using @Lorak-mmk's reproducer that disabling the feature solves the problem.

Fixes scylladb/scylladb#15530

### CODE REVIEW

**Comment by kbr-scylla** [comment]
Regarding how https://github.com/scylladb/scylladb/commit/3d4398d1b2b9bc2139718e49a806214171ee081f could break stuff -- I don't know. But maybe it works something like this: with the hash method, when we alter type definition that a given table refers to, the hash of this table changes (? -- need to verify). On the other hand, with group 0 schema versioning, the version of the table is constant, even though the types that it refers to may change. (Only altering the table directly may change its version.) And somehow not changing the version breaks the schema object too? Maybe we use the version to obtain the schema object from cache or something, and since the version is the same, we obtain the old `schema_ptr` which refers to old type definitions. @tgrabiec does this seem plausible?

**Comment by kbr-scylla** [comment]
```
cqlsh> create keyspace ks with replication = {'class': 'SimpleStrategy', 'replication_factor': 1};

Warnings :
Using Replication Factor replication_factor=1 lower than the minimum_replication_factor_warn_threshold=3 is not recommended.

cqlsh> create type ks.typ (a int, b int);
cqlsh> create table ks.t (pk int primary key, v typ);
cqlsh> alter type ks.typ add c int;

```

the `alter` causes:
```
INFO  2023-10-10 16:30:52,022 [shard 0:stat] migration_manager - Prepare Update User Type: typ
INFO  2023-10-10 16:30:52,022 [shard 0:stat] migration_manager - Update table 'ks.t' From org.apache.cassandra.config.CFMetaData@0x6000007a9500[cfId=9a2da350-6779-11ee-bcd1-980a508b85ed,ksName=ks,cfName=t,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type,org.apache.cassandra.db.marshal.ColumnToCollectionType(76:org.apache.cassandra.db.marshal.UserType(ks,747970,61:org.apache.cassandra.db.marshal.Int32Type,62:org.apache.cassandra.db.marshal.Int32Type))),comment=,readRepairChance=0,dcLocalReadRepairChance=0,tombstoneGcOptions={"mode":"timeout","propagation_delay_in_seconds":"3600"},gcGraceSeconds=864000,keyValidator=org.apache.cassandra.db.marshal.Int32Type,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=pk, type=org.apache.cassandra.db.marshal.Int32Type, kind=PARTITION_KEY, componentIndex=0, droppedAt=-9223372036854775808}, ColumnDefinition{name=v, type=org.apache.cassandra.db.marshal.UserType(ks,747970,61:org.apache.cassandra.db.marshal.Int32Type,62:org.apache.cassandra.db.marshal.Int32Type), kind=REGULAR, componentIndex=null, droppedAt=-9223372036854775808}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={enabled=true},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL","rows_per_partition":"ALL"},cdc={},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,triggers=[],isDense=false,version=006a264c-194e-3376-a385-cdfe21fdd1c1,droppedColumns={},collections={76 : org.apache.cassandra.db.marshal.UserType(ks,747970,61:org.apache.cassandra.db.marshal.Int32Type,62:org.apache.cassandra.db.marshal.Int32Type)},indices={}] To org.apache.cassandra.config.CFMetaData@0x6000007a7180[cfId=9a2da350-6779-11ee-bcd1-980a508b85ed,ksName=ks,cfName=t,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type,org.apache.cassandra.db.marshal.ColumnToCollectionType(76:org.apache.cassandra.db.marshal.UserType(ks,747970,61:org.apache.cassandra.db.marshal.Int32Type,62:org.apache.cassandra.db.marshal.Int32Type,63:org.apache.cassandra.db.marshal.Int32Type))),comment=,readRepairChance=0,dcLocalReadRepairChance=0,tombstoneGcOptions={"mode":"timeout","propagation_delay_in_seconds":"3600"},gcGraceSeconds=864000,keyValidator=org.apache.cassandra.db.marshal.Int32Type,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=pk, type=org.apache.cassandra.db.marshal.Int32Type, kind=PARTITION_KEY, componentIndex=0, droppedAt=-9223372036854775808}, ColumnDefinition{name=v, type=org.apache.cassandra.db.marshal.UserType(ks,747970,61:org.apache.cassandra.db.marshal.Int32Type,62:org.apache.cassandra.db.marshal.Int32Type,63:org.apache.cassandra.db.marshal.Int32Type), kind=REGULAR, componentIndex=null, droppedAt=-9223372036854775808}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={enabled=true},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL","rows_per_partition":"ALL"},cdc={},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,triggers=[],isDense=false,version=9d251160-6779-11ee-bcd1-980a508b85ed,droppedColumns={},collections={76 : org.apache.cassandra.db.marshal.UserType(ks,747970,61:org.apache.cassandra.db.marshal.Int32Type,62:org.apache.cassandra.db.marshal.Int32Type,63:org.apache.cassandra.db.marshal.Int32Type)},indices={}]
INFO  2023-10-10 16:30:52,030 [shard 0:stre] schema_tables - Altering ks.t id=9a2da350-6779-11ee-bcd1-980a508b85ed version=641c74d4-b6e5-3b90-ab2e-cce456312459
INFO  2023-10-10 16:30:52,039 [shard 0:stre] query_processor - Column definitions for ks.t changed, invalidating related prepared statements
INFO  2023-10-10 16:30:52,040 [shard 1:stre] query_processor - Column definitions for ks.t changed, invalidating related prepared statements
INFO  2023-10-10 16:30:52,041 [shard 2:stre] query_processor - Column definitions for ks.t changed, invalidating related prepared statements
INFO  2023-10-10 16:30:52,059 [shard 0:stre] schema_tables - Schema version changed to 66d598e9-4b7e-32d6-aa0d-90f3832df746
```


In particular:
```
Altering ks.t id=9a2da350-6779-11ee-bcd1-980a508b85ed version=641c74d4-b6e5-3b90-ab2e-cce456312459
```
Then another alter:
```
INFO  2023-10-10 16:31:47,229 [shard 0:stre] schema_tables - Altering ks.t id=9a2da350-6779-11ee-bcd1-980a508b85ed version=2b476e56-af6d-3634-a4d2-08ab41fc796e
```

so the table version changes when we alter a type that the table definition depends on.

BUT I verified that this is happening both with and without group 0 schema versioning.

Since this is happening with group 0 schema versioning, apparently the schema change coordinator is also constructing a new `schema` object for that table and using it to create mutations that alter the table.

Maybe there's a race on the side of schema change coordinator -- maybe there's unlikely chance that the coordinator would get the same `schema` object, with the same version? And then it would lead to committing a command with broken schema?

**Comment by avikivity** [comment]
I think we revert it instead, rather than leaving dead and broken code in there.

**Comment by kbr-scylla** [comment]
> I think we revert it instead, rather than leaving dead and broken code in there.

It could be that the code is not broken, but exposes a pre-existing problem. I want to get to the bottom of it and fix the root cause, then reenable group 0 schema versioning.

**Comment by kbr-scylla** [comment]
> Maybe there's a race on the side of schema change coordinator -- maybe there's unlikely chance that the coordinator would get the same schema object, with the same version? And then it would lead to committing a command with broken schema?

This is indeed what seems to be happening.

Using @Lorak-mmk's reproducer I reproduced the problem again, and looked at schemas that are being committed by group 0.
When the problem reproduces, indeed the table version does not change:

```
// previous alter
INFO  2023-10-10 16:40:50,617 [shard  0:stre] schema_tables - Altering ks.entity id=ff76f080-677a-11ee-9054-b0ee49ce148b version=fff150a0-677a-11ee-9054-b0ee49ce148b                                                                                         

...
// then the breaking user type update
INFO  2023-10-10 16:40:50,832 [shard  0:stat] migration_manager - Prepare Update User Type: user_refs                                                                                                                                                         
...
INFO  2023-10-10 16:40:51,409 [shard  0:stre] schema_tables - Altering ks.entity id=ff76f080-677a-11ee-9054-b0ee49ce148b version=fff150a0-677a-11ee-9054-b0ee49ce148b                                                                                         
```

So the schema change coordinator appears to be using old `schema` object when doing the alter type on the failing iteration.

In contrast, a non-failing iteration changes the version at that step.

**Comment by kbr-scylla** [comment]
Ahhhhh, uuid vs timeuuid comparators strike again...
In the past 3 months I had to deal with this problem like 3 times already!

tl;dr: type update triggers view update and base table update, both generate mutations for `system_schema.scylla_tables`, the view updates use old version, the base table update uses new version, but they are sent in a single schema change command using the same timestamp, and the base table update's version wins because it's greater when using UUID comparison, and `version` column in `system_schema.scylla_tables` is of type `uuid`.

---------

Some context first.

When we call `schema_builder::build()`, if we didn't call `with_version` before, `build()` will generate a version for this schema object as follows:
```cpp
        new_raw._version = table_schema_version(utils::UUID_gen::get_time_UUID());
```

This version is then sent as part of schema change command inside `system_schema.scylla_tables` mutation.
This is old behavior, before 3d4398d1b2b9bc2139718e49a806214171ee081f.

With group 0 schema versioning disabled, this version would be ignored during schema mutation apply step (it would be removed from the mutation), and a new version based on hash would be generated. What 3d4398d1b2b9bc2139718e49a806214171ee081f changed, is that we no longer ignore this version, but let it get persisted in `system_schema.scylla_tables` and then use it as the table's version.

The fact that it's a timeuuid doesn't matter when there's a single `system_schema.scylla_tables` mutation inside a schema change command. For commands that eventually get applied successfully, thanks to group 0's state_id mechanism, the timestamps of mutations present there are increasing. So the version of a  subsequent schema change will win over the version of previous schema change.

But the situation becomes more interesting when there are multiple `system_schema.scylla_tables` mutations inside a schema change command, and these mutations all use the same timestamp, and there are multiple mutations for the same row in `scylla_tables` (!) which modify the version column.

Now it turns out that such scenario happens when we have materialized views attached to the table. Our schema change command then contains multiple conflicting `system_schema.scylla_tables` mutations, and they have the same timestamp, so conflict resolution is done by comparing the values!

I printed what mutations are generated for the schema change commands, and here's what happens when we update one of the user defined types:
```
INFO  2023-10-10 17:38:06,418 [shard  0:stat] migration_manager - Prepare Update User Type: obs_entity
INFO  2023-10-10 17:38:06,433 [shard  0:stat] database - ks.entity_by_unique_id build using newly generated version 01d75510-6783-11ee-9f6b-f65596552528
INFO  2023-10-10 17:38:06,433 [shard  0:stat] migration_manager - Update view 'ks.entity_by_unique_id
INFO  2023-10-10 17:38:06,434 [shard  0:stat] schema_tables - make scylla tables mutation ks.entity table version ff5a6e30-6782-11ee-9f6b-f65596552528 timestamp 1696952286417867
INFO  2023-10-10 17:38:06,443 [shard  0:stat] database - ks.entity_by_type build using newly generated version 01d8dbb0-6783-11ee-9f6b-f65596552528
INFO  2023-10-10 17:38:06,443 [shard  0:stat] migration_manager - Update view 'ks.entity_by_type
INFO  2023-10-10 17:38:06,444 [shard  0:stat] schema_tables - make scylla tables mutation ks.entity table version ff5a6e30-6782-11ee-9f6b-f65596552528 timestamp 1696952286417867
INFO  2023-10-10 17:38:06,453 [shard  0:stat] database - ks.entity build using newly generated version 01da6250-6783-11ee-9f6b-f65596552528
INFO  2023-10-10 17:38:06,454 [shard  0:stat] migration_manager - Update table 'ks.entity
INFO  2023-10-10 17:38:06,454 [shard  0:stat] schema_tables - make scylla tables mutation ks.entity table version 01da6250-6783-11ee-9f6b-f65596552528 timestamp 1696952286417867
INFO  2023-10-10 17:38:07,306 [shard  0:stre] schema_tables - Altering ks.entity id=fef9feb0-6782-11ee-9f6b-f65596552528 version=ff5a6e30-6782-11ee-9f6b-f65596552528
```

Updating the user type caused the following:
- the `ks.entity_by_unique_id` view gets updated, which also generates a `system_schema.scylla_tables` mutation for the base table `ks_entity`
- same for the `ks.entity_by_type` view
- and the base table `ks.entity` itself gets updated which also generates a mutation for it

so we have 3 mutations for `ks.entity`.
And the mutations coming from views are using the old version.
The mutation coming from base table update is using a newly generated version.

First view gives:
```
make scylla tables mutation ks.entity table version ff5a6e30-6782-11ee-9f6b-f65596552528 timestamp 1696952286417867
```
Second view gives:
```
make scylla tables mutation ks.entity table version ff5a6e30-6782-11ee-9f6b-f65596552528 timestamp 1696952286417867
```
Base table gives:
```
make scylla tables mutation ks.entity table version 01da6250-6783-11ee-9f6b-f65596552528 timestamp 1696952286417867
```

And now which version will be the resulting version after all the mutations get written to `system_schema.scylla_tables`?
The greatest one.
And if we were lucky, the `version` column would have been a `timeuuid` type. But it's a `uuid` type. So `ff5a6e30-6782-11ee-9f6b-f65596552528` is actually greater than `01da6250-6783-11ee-9f6b-f65596552528`.

And therefore, even though we update the table, the version ends up being the old one...

**Comment by avikivity** [comment]
Ouch this is painful.

**Comment by avikivity** [comment]
How do we solve it? We can't change the type. Do we implement a custom conflict resolver? sounds like a lot of work that will bring its own bugs.

**Comment by kbr-scylla** [comment]
I think we can simply skip the base table update mutations that are generated by view updates.

So alter type does this:
```cpp
future<std::vector<mutation>> alter_type_statement::prepare_announcement_mutations(service::storage_proxy& sp, api::timestamp_type ts) const {
...
        if (modified) {
            if (schema->is_view()) {
                auto res = co_await service::prepare_view_update_announcement(sp, view_ptr(cfm.build()), ts);
                std::move(res.begin(), res.end(), std::back_inserter(m));
            } else {
                auto res = co_await service::prepare_column_family_update_announcement(sp, cfm.build(), false, {}, ts);
                std::move(res.begin(), res.end(), std::back_inserter(m));
            }
        }
```

The troublesome mutations, that provide the old version for the base table, come from `prepare_view_update_announcement`.

```cpp
future<std::vector<mutation>> prepare_view_update_announcement(storage_proxy& sp, view_ptr view, api::timestamp_type ts) {
...
        auto mutations = db::schema_tables::make_update_view_mutations(keyspace, view_ptr(old_view), std::move(view), ts, true);
...
```

the interesting part is that last parameter, set to `true`. This is `bool include_base`:
```cpp
std::vector<mutation> make_update_view_mutations(lw_shared_ptr<keyspace_metadata> keyspace,
                                                 view_ptr old_view,
                                                 view_ptr new_view,
                                                 api::timestamp_type timestamp,
                                                 bool include_base)
{
    std::vector<mutation> mutations;
    if (include_base) {
        // Include the serialized base table mutations in case the target node is missing them.
        auto base = keyspace->cf_meta_data().at(new_view->view_info()->base_name());
        add_table_or_view_to_schema_mutation(base, timestamp, true, mutations);
    }
```

In Raft mode, I think we can simply set `include_base = false`. This comment does not apply in Raft mode:
```
 // Include the serialized base table mutations in case the target node is missing them.
```

In fact, in the case of type updates, I wonder if it's necessary at all. We could modify `prepare_view_update_announcement` to take `include_base`, and inside `alter_type_statement::prepare_announcement_mutations` set it to `false` -- Raft mode or not. I think this would be even simpler, no need to check for being in Raft mode (which is not that trivial).

**Comment by kbr-scylla** [comment]
Including these mutations would also not be a problem if we included the actual old mutations -- i.e. the ones that are present on disk at this point. These mutations have the smaller timestamp because they were generated in an older schema change command.

The problem arises because this code is generating **new** mutations for the base table, with the new timestamp, but with old schema version.

**Comment by kbr-scylla** [comment]
We could also do this...
```diff
diff --git a/db/schema_tables.cc b/db/schema_tables.cc
index b22faa1ec9..b9ca6e57a0 100644
--- a/db/schema_tables.cc
+++ b/db/schema_tables.cc
@@ -3515,7 +3515,7 @@ std::vector<mutation> make_update_view_mutations(lw_shared_ptr<keyspace_metadata
     if (include_base) {
         // Include the serialized base table mutations in case the target node is missing them.
         auto base = keyspace->cf_meta_data().at(new_view->view_info()->base_name());
-        add_table_or_view_to_schema_mutation(base, timestamp, true, mutations);
+        add_table_or_view_to_schema_mutation(base, timestamp-1, true, mutations);
     }
     add_table_or_view_to_schema_mutation(new_view, timestamp, false, mutations);
     make_update_columns_mutations(old_view, new_view, timestamp, false, mutations);
```

that's the most lazy solution I think.

Although need to think if it won't have some weird unintended consequences.

But at a first glance it shouldn't.
In non-Raft mode concurrent schema changes don't work anyway (this code is not really fixing all the problems with concurrent schema changes).
In Raft mode, `timestamp-1` will be greater or equal than the timestamp from previous commands. If it's greater, then the mutations will override mutations from previous commands, as before. If it's equal, then we will solve conflicts on a per-cell basis based on value, but unless there's some other bug, all values should be the same.

But thinking about this, the fact that we include these base mutations is only making stuff unnecessarily more complicated in Raft mode. Maybe we should just get rid of them to simplify reasoning. This would require checking if we're in Raft mode, which will require plumbing some booleans through cql statement interfaces, but that's a sacrifice I'm willing to make.

**Comment by avikivity** [comment]
There's something with a delete timestamp being a value timestamp minus 1. Better to use the original timestamp.

**Comment by tgrabiec** [comment]
> There's something with a delete timestamp being a value timestamp minus 1. Better to use the original timestamp.

Yes, in `make_table_deleting_mutations()` we emit such tombstones, will be part of `make_create_table_mutations()`.

**Comment by bhalevy** [comment]
I haven't testedt it, but if we would use UUIDv7 instead of UUIDv1 for the version they should sort properly as plain UUID.

Cc @avikivity @nyh

**Comment by avikivity** [comment]
> I haven't testedt it, but if we would use UUIDv7 instead of UUIDv1 for the version they should sort properly as plain UUID.
> 
> Cc @avikivity @nyh @tchaikov

Let's not mix in yet another thing.

**Comment by avikivity** [comment]
@kbr-scylla CI fails, looks like relevant failures. I'll revert the commit.

**Comment by kbr-scylla** [comment]
> @kbr-scylla CI fails, looks like relevant failures. I'll revert the commit.

That's because I forgot to disable my own tests that were testing the versioning feature (－‸ლ) (of course they fail when I disable the feature).

**Comment by avikivity** [comment]
Yes, I figured, but wanted to clear the nasty bug as quickly as I could.

**Comment by kbr-scylla** [comment]
I started working on fixing the root cause, sent https://github.com/scylladb/scylladb/pull/15695 as part of the effort

**Comment by mykaul** [comment]
@kbr-scylla - will this be completed before your vacation? If it's critical, either get it in or delegate it to someone.

**Comment by kbr-scylla** [comment]
This PR is no longer relevant.
